include/envoy/server/options.h

```diff
@@ -78,6 +79,11 @@ class Options {
    */
   virtual const std::string& configYaml() const PURE;
 
+  /**
+   * @return const envoy::config::bootstrap::v2::Bootstrap& a botstrap proto object.
+   */
+  virtual const envoy::config::bootstrap::v2::Bootstrap& configProto() const PURE;
```

Comment:
Also can we make it clear is this only the config that's merged into a bootstrap after merging config_path+config_yaml?

It may worthwhile to have a comment denoting this is not* the "full" bootstrap config that the OptionsImpl contains. And have a "bootstrap()" method which returns the "merged" bootstrap would be nice.

Reply:
As far as I know, only https://github.com/envoyproxy/envoy/blob/ace23e9de99d023cb797405ddb02dc28df2d916a/source/server/server.h#L268 holds the final `Bootstrap` proto object after all the merges. Each of `configPath`, `configYaml`, and now `configProto` only hold what was explicitly provided to them, *not* the merged version of them.

Reply:
It might be better to move the logic of composing bootstrap proto from server.cc into the options impl(together with some validation), as that's essentially where all the bootstrap config comes from. Server only need to ask for Options::bootstrap() to get the bootstrap. That way Envoy can be run on some diskless machine easily, and we will have a single source of truth for bootstrap (note server also holds a copy of options), eliminates any possible confusion.

Reply:
In my mind the `Options` class is used as a sort of temporary object holding desired configuration flags. It is then processed in `server.cc` to produce the final state of the server.

If we wanted some way to interrogate the final configuration of the server, I'd think it should be returned from `Server::InstanceImpl` and not the `Options` class.

Reply:
No strong opinion on where to get the final bootstrap. 
Server instanceImpl (at present) gets all bootstrap config from Options. 
The difficulty of getting the final bootstrap "outside of " Envoy is not resolved in this PR tho. Since we can only get the bootstrap via Server instance which is not exposed by mainCommon.


Please comment here that bootstrap proto is the last one merged into the final bootstrap config.

Reply:
Gotcha, yeah I think if we need to retrieve the final bootstrap we should do it via the `Server` object.

Added comment clarifying merge order of `configProto`.